### ISSUE TITLE
Segmentation fault even for small file and huge amount of memory

### ISSUE THREAD

**Issue by vicbp1** [YingZhou001/IBDkin#3]
Hi Ying
I am trying to run IBDkin in a data set of 85K individuals (2.8M SNPs). The average size of each compressed hapibd output is 50GB. I am running the following line in a machine with 500Gb

for i in $(seq 1 5); do ${IBDkin} --ibdfile ibd.txt --map geneticmap.txt --ind complete_list_individuals_freeze1 --range ranges.txt --nthreads 60 --out kinship_analysis_part${i} --outmask --outcoverage --kinship 0.03125 --part 5 ${i} ; done

Even when I run just for the small chr (~27Gb) I got the Segmentation fault error when ibdkin is reading the hapibd file.

First pass ...
	reading /21_split1.ibd.gz
Segmentation fault (core dumped)

**Do you have any recommendation?**

Another question, does ibdkin have any bias related to the kinship coeffient inferences for recent admixed individuals?

**Comment by YingZhou001** [comment]
Hi  Victor,

Thanks for reaching.
Based on your situation, I will try to add option "--part 10 1" to your command. If the problem is still there, I will check the input files.

Best,
Ying

**Comment by vicbp1** [comment]
Thank for your answer!
I double checked the inputs and they look ok. For values like --part 10, 20, or 50 the result is the same even when I increase the memory availability (700G). Just for one chr that I had a different error : Error!, maker ranges for chromosome  do not cover the position 21138926

Even when my data have that region

Any thoughts?

**Comment by YingZhou001** [comment]
Hi Victor, I don't think this is a memory problem. 
Did you check the range file? it should cover all your markers. 
And also check the chromosome identifier, it should be in the same codes for all input files.
If the range file is good, I will cehck 1) ibd file, compare the format with the example to look difference. 2) for each IBD segment, the end position should be larger than the start position. 3) the genetic map, both the physical and genetic positions should be in increasing order. 


Best,
Ying

**Comment by vicbp1** [comment]
Hi Ying, 

I found that there was something wrong during the compression of the files and now is working
Thank you so much for your time!

**Comment by vicbp1** [comment]
Hi Ying,

After checking all my files and running again I have found this error:

**Error!, maker ranges for chromosome  do not cover the position 4160868**

What it means exactly? Do I have to modify my range sfile?

**Comment by YingZhou001** [comment]
Hi Victor,

Yes, please try to modify the range file.
The range file is used for allocate memory so that it can save time for
ibdkin for reading ibd file.

Best,
Ying

**Comment by bguo068** [comment]
Hi @YingZhou001, 

It has been a while since we met via Zoom in the MCMC workshop this summer :).
Victor and I are from the same group. We looked at this issue together this afternoon. We figured that the issue was caused by the memory overflow from `str1` or `str2` to `chr`.
https://github.com/YingZhou001/IBDkin/blob/61ccecbf3cae8b5dd4b42716aeed099e35a3c3ed/src-v2.8.7.7/parallel.c#L623-L625

https://github.com/YingZhou001/IBDkin/blob/61ccecbf3cae8b5dd4b42716aeed099e35a3c3ed/src-v2.8.7.7/parallel.c#L634-L636

It would not be an issue if the sample names are short. But we happened to have long sample names in our IBD input files (as long as 58 bytes, longer than buffer size (BUFF1  or 32) .
https://github.com/YingZhou001/IBDkin/blob/61ccecbf3cae8b5dd4b42716aeed099e35a3c3ed/src-v2.8.7.7/head.h#L16-L17

After we edited `#define BUFF1 32 ` to `#define BUFF1 128`, `#define BUFF 128 ` to `#define BUFF 512` and recompiled the code, it runs without crashing immediately. Hope it runs smoothly to the end. @vicbp1
